### PR TITLE
fix: merge compatible definitions in union types

### PR DESCRIPTION
**Pull request by dmchurch**
Most validation keywords apply to only one of the basic types, so a
"string" type and an "array" type can share a definition without
colliding as a ["string", "array"] type, as long as they don't have any
incompatibilities with each other. Modify UnionTypeFormatter to collapse
these disjoint types into a single definition, without using anyOf.

### CODE REVIEW

**Comment by dmchurch** [comment]
I've classed this as a "fix" because it doesn't change the semantic meaning of the output, and it doesn't enable new features for schema developers. Feel free to upgrade it to a "feat", though.

File: src/Utils/mergeDefinitions.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Returns the merged definition, or null if the two defs were not disjoint.
+ */
+export function mergeDefinitions(def1: Definition, def2: Definition): Definition | null {
+    const { type: type1, ...props1 } = def1;
+    const { type: type2, ...props2 } = def2;
+    const types = [type1!, type2!].flat();
+    if (!type1 || !type2 || uniqueArray(types).length !== types.length) {
+        return null;
+    }
+    const keys = [Object.keys(props1), Object.keys(props2)].flat();
```
Comment (domoritz) [inline src/Utils/mergeDefinitions.ts:20]:
```suggestion
    const keys = [...Object.keys(props1), ...Object.keys(props2)];
```

**Review by domoritz — COMMENTED** [review]
Thanks for the pull request. Please take a look at the tests.

**Comment by domoritz** [comment]
I think you probably just need to update the Vega-Lite test case. You can use the provided script to do it automatically.

**Comment by dmchurch** [comment]
Right! I'll get on the tests now. Some of them weren't passing even in my original checkout, but at the very least I can make sure that anything affected by this change gets updated.

**Comment by domoritz** [comment]
Try `yarn test:update`

**Comment by dmchurch** [comment]
I've actually got a couple test cases I want to add to exercise some of the weirder edge cases involving definition collapse, but I'm setting this down for the night. In any event, the existing test cases already exercise it fairly well.

**Comment by dmchurch** [comment]
Whew! That got _much_ bigger than expected. I added some schema validation tests, which uncovered some issues with the existing code, which required a more elaborate test harness, which means the test harness needs more tests... you get the idea. One of these days I'll get all those yaks shaved.

**Comment by dmchurch** [comment]
...and I literally just now realized that the makeExemplar code ought to be living on-or-near the BaseType classes themselves, which would clean up the huge mess of if-else-if-else in makeExemplar.ts. Oh well, there are more revisions I need to make anyway...

Feel free to offer critique/direction if this isn't going in a useful direction, btw, I'm not averse to scrapping things and rebuilding.

**Comment by dmchurch** [comment]
...and mergeDefinitions would get so very much smaller if I just called it twice, once in each direction. Yeah, don't merge this yet, I've got work to do.

**Comment by domoritz** [comment]
Thank you for working on this and making sure the feature is clean. I definitely think this is a good improvement.

File: src/Utils/removeUnreachable.ts [hunk 2]
```diff
@@ -20,39 +30,23 @@ function addReachable(
...
-        for (const def of definition.allOf) {
-            addReachable(def, definitions, reachable);
-        }
-    } else if (definition.oneOf) {
-        for (const def of definition.oneOf) {
-            addReachable(def, definitions, reachable);
-        }
-    } else if (definition.not) {
+    } else {
+        addReachable(definition.anyOf, definitions, reachable);
```
Comment (domoritz) [inline src/Utils/removeUnreachable.ts:34]: This looks great

**Review by domoritz — COMMENTED** [review]
Overall, this looks great. Please wrap up the remaining issues and mark the pull request as ready for (final) review.

File: test/utils.ts [hunk 3]
```diff
@@ -8,16 +8,88 @@ import { createFormatter } from "../factory/formatter";
...
 addFormats(validator);
 
 const basePath = "test/valid-data";
 
-export function createGenerator(config: Config): SchemaGenerator {
-    const program: ts.Program = createProgram(config);
-    return new SchemaGenerator(program, createParser(program, config), createFormatter(config), config);
+interface Validator<T> {
+    (value: T): boolean | Promise<unknown>;
+    //                    ^^^^^^^^^^^^^^^^ *Why* does validateSchema() sometimes return this? Who knows! >.<
```
[the remark this answers is not in the document]
Comment (dmchurch) [reply test/utils.ts:24]: Haha, that's not us! Ajv has a very unfortunate coding error, the .d.ts file has a different signature from the online documentation. This is just working around that.

I suppose I could open an issue on their repo, though.

File: test/vega-lite/schema.json [hunk 4]
```diff
@@ -1558,18 +1554,14 @@
...
+          "description": "A string or array of strings indicating the name of custom styles to apply to the axis. A style is a named collection of axis property defined within the [style configuration](https://vega.github.io/vega-lite/docs/mark.html#style-config). If style is an array, later styles will override earlier styles.\n\n__Default value:__ (none) __Note:__ Any specified style will augment the default style. For example, an x-axis mark with `\"style\": \"foo\"` will use `config.axisX` and `config.style.foo` (the specified style `\"foo\"` has higher precedence).",
+          "items": {
+            "type": "string"
+          },
+          "type": [
```
[the remark this answers is not in the document]
Comment (dmchurch) [reply test/vega-lite/schema.json:1561]: This is as intended/expected, actually! The type-specific JSON schema validation constraints specifically only apply when the value is of the appropriate type, which is why a lot of definitions can get collapsed together. If you included "items" without "type" it would mean "The value can be any type, but if it's an array its elements must be strings". So, this schema means "The value can be an array or a string. If the value is an array, its elements must be strings".

That's why all the type-specific properties have different names - "minLength" only applies to strings, but the same concept applied to arrays is "minItems".

File: src/Utils/makeExemplar.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+import { ReferenceType } from "../Type/ReferenceType";
+import { RestType } from "../Type/RestType";
+import { StringType } from "../Type/StringType";
+import { SymbolType } from "../Type/SymbolType";
+import { TupleType } from "../Type/TupleType";
+import { UndefinedType } from "../Type/UndefinedType";
+import { UnionType } from "../Type/UnionType";
+
+export function makeExemplar(type: BaseType | undefined): unknown {
+    return makeExemplars(type)[0];
```
[the remark this answers is not in the document]
Comment (dmchurch) [reply src/Utils/makeExemplar.ts:23]: Yeah, I definitely need to refactor this into something cleaner. On my to-do list, certainly.

File: test/vega-lite/schema.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (domoritz) [reply test/vega-lite/schema.json:1561]: I see. As long as we can still use the schema to generate APIs from it (which we do), I'm okay.

File: test/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (domoritz) [reply test/utils.ts:24]: That would be great and then add a link to it in the code. Thanks!

**Comment by domoritz** [comment]
@dmchurch do you plan to wrap this up soon?

**Comment by domoritz** [comment]
@dmchurch can you please finish up this pull request?
